### PR TITLE
[BUG][python] Support named arrays

### PR DESCRIPTION
**Pull request by jirikuncar**
**The model for named array is not generated.**

### PR checklist

- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [ ] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [ ] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

File: samples/openapi3/client/petstore/python-experimental/petstore_api/api/fake_api.py [hunk 1]
```diff
@@ -600,6 +601,109 @@ def __fake_outer_string_serialize(
...
+            )
+            kwargs['_check_return_type'] = kwargs.get(
+                '_check_return_type', True
+            )
+            kwargs['_host_index'] = kwargs.get('_host_index', 0)
+            return self.call_with_http_info(**kwargs)
+
+        self.get_array_of_enums = Endpoint(
+            settings={
+                'response_type': (array_of_enums.ArrayOfEnums,),
```
Comment (jirikuncar) [inline samples/openapi3/client/petstore/python-experimental/petstore_api/api/fake_api.py:665]: ❗

File: samples/openapi3/client/petstore/python-experimental/petstore_api/api/fake_api.py [hunk 2]
```diff
@@ -36,6 +36,7 @@
 )
 from petstore_api.models import health_check_result
 from petstore_api.models import outer_composite
+from petstore_api.models import array_of_enums
```
Comment (jirikuncar) [inline samples/openapi3/client/petstore/python-experimental/petstore_api/api/fake_api.py:39]: ❗

File: modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml [hunk 3]
```diff
@@ -1703,6 +1716,10 @@ components:
         - placed
         - approved
         - delivered
+    ArrayOfEnums:
+      type: array
+      items:
+        $ref: '#/components/schemas/OuterEnum'
```
Comment (spacether) [inline modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1722]: Does the model AnimalFarm already demonstrate your use case?

**Comment by spacether** [comment]
My investigation shows that this line:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L420
Is preventing these models from being generated
because the array model does not have any validations or enums.
A couple of solutions that come to mind here are:
1. define the data_type in this case as `[outer_enum. OuterEnum]`. This is probably the simpler solution to get working.
2. uncomment the suppression of these models and ensure that these models are generated correctly. When I do this is see that the openapi_types data in the produced models is incorrect. Also the produced models are subclases of the ModelNormal class when they should be subclasses of the ModelSimple class

What do you think about each of these possible solutions?

File: modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml (hunk 3 above)
Comment (jirikuncar) [reply modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1722]: I haven't seen it used as top level object. I can use it here instead. WDYT?

Comment (spacether) [reply modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1722]: What do you mean by top level object? Do you mean used in an endpoint?
Testing shows that the same code path is hit with that model and this one so if you can use that schema and avoid adding this schema it would be preferred because it will keep our sample spec smaller.

Comment (jirikuncar) [reply modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1722]: As I was not sure about the purpose, I have initially created new model. I am using the `AnimalFarm` now.

**Comment by jirikuncar** [comment]
I have added `FarmApi` with `/farm/animals` endpoint returning array of `Animal`s.

File: samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# animal_farm.AnimalFarm
+
+## Properties
+Name | Type | Description | Notes
+------------ | ------------- | ------------- | -------------
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md:5]: Can you fix the doc template so it lists the information for the AnimalFarm model here?
Name: N/A
Type: [animal.Animal]

File: modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml [hunk 5]
```diff
@@ -1097,6 +1097,19 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/HealthCheckResult'
+  /farm/animals:
+    get:
+      tags:
+        - farm
```
Comment (spacether) [inline modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1103]: This tag has generated a new api which adds another file.
Can you change it to one of the existing tags like `pet` or `fake`
That way a new file will not be generated

File: samples/openapi3/client/petstore/python-experimental/petstore_api/api/farm_api.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
+# coding: utf-8
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/petstore_api/api/farm_api.py:1]: Can you delete this file after you switch the api tag to put the farm endpoint into the fake api?

File: samples/openapi3/client/petstore/python-experimental/test/test_animal_farm.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        pass
+
+    def tearDown(self):
+        pass
+
+    def testAnimalFarm(self):
+        """Test AnimalFarm"""
+        # FIXME: construct object with mandatory attributes with example values
+        # model = AnimalFarm()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_animal_farm.py:39]: How about adding a test here like
animal = [animal.Animal(blah)]
model = AnimalFarm(animal)
assert model[0] === animal

- How about adding a test in test/test_fake_api.py where we send an instance of AnimalFarm to the endpoint that accepts it?

- How about adding a deserialization test like we have [here](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/openapi3/client/petstore/python-experimental/tests/test_deserialization.py#L86) to show that our code can receive back array models and deserialize them correctly?

File: samples/openapi3/client/petstore/python-experimental/test/test_drawing.py [hunk 8]
```diff
@@ -42,130 +47,12 @@ def setUp(self):
...
-        Validate the scenario when the type of a OAS property is 'oneOf', and the 'oneOf'
-        schema is specified as a reference ($ref), not an inline 'oneOf' schema.
-        Further, the 'oneOf' schema has the 'nullable' attribute (as introduced in
-        OpenAPI 3.0 and deprecated in 3.1).
-        """
+    def testDrawing(self):
+        """Test Drawing"""
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Drawing()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_drawing.py:54]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_fruit.py [hunk 9]
```diff
@@ -39,184 +39,10 @@ def tearDown(self):
 
     def testFruit(self):
         """Test Fruit"""
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Fruit()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_fruit.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_fruit_req.py [hunk 10]
```diff
@@ -39,142 +39,10 @@ def tearDown(self):
 
     def testFruitReq(self):
         """Test FruitReq"""
+        # FIXME: construct object with mandatory attributes with example values
+        # model = FruitReq()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_fruit_req.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_gm_fruit.py [hunk 11]
```diff
@@ -39,193 +39,10 @@ def tearDown(self):
 
     def testGmFruit(self):
         """Test GmFruit"""
+        # FIXME: construct object with mandatory attributes with example values
+        # model = GmFruit()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_gm_fruit.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_mammal.py [hunk 12]
```diff
@@ -39,11 +44,9 @@ def tearDown(self):
...
     def testMammal(self):
         """Test Mammal"""
-
-        # tests that we can make a BasquePig by traveling through descendant discriminator in Pig
-        model = Mammal(class_name="BasquePig")
-        from petstore_api.model import basque_pig
-        assert isinstance(model, basque_pig.BasquePig)
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Mammal()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_mammal.py:49]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_outer_enum.py [hunk 13]
```diff
@@ -29,16 +29,9 @@ def tearDown(self):
...
-        self.assertIsNone(inst)
-
-        inst = OuterEnum('approved')
-        assert isinstance(inst, OuterEnum)
-
-        with self.assertRaises(petstore_api.ApiValueError):
-            OuterEnum('garbage')
+        # FIXME: construct object with mandatory attributes with example values
+        # model = OuterEnum()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_outer_enum.py:34]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_parent_pet.py [hunk 14]
```diff
@@ -39,12 +39,9 @@ def tearDown(self):
...
         """Test ParentPet"""
-
-        # test that we can make a ParentPet from a ParentPet
-        # which requires that we travel back through ParentPet's allOf descendant
-        # GrandparentAnimal, and we use the descendant's discriminator to make ParentPet
-        model = ParentPet(pet_type="ParentPet")
-        assert isinstance(model, ParentPet)
+        # FIXME: construct object with mandatory attributes with example values
+        # model = ParentPet()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_parent_pet.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_quadrilateral.py [hunk 15]
```diff
@@ -39,10 +39,9 @@ def tearDown(self):
 
     def testQuadrilateral(self):
         """Test Quadrilateral"""
-        instance = Quadrilateral(shape_type="Quadrilateral", quadrilateral_type="ComplexQuadrilateral")
-        assert isinstance(instance, complex_quadrilateral.ComplexQuadrilateral)
-        instance = Quadrilateral(shape_type="Quadrilateral", quadrilateral_type="SimpleQuadrilateral")
-        assert isinstance(instance, simple_quadrilateral.SimpleQuadrilateral)
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Quadrilateral()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_quadrilateral.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_shape.py [hunk 16]
```diff
@@ -39,71 +39,10 @@ def tearDown(self):
...
-                   "property '{}'='{}' at path: ()"
-                   )
-        with self.assertRaisesRegexp(
-                petstore_api.ApiValueError,
-                err_msg.format("shapeType", "Circle")
-        ):
-            Shape(shape_type="Circle")
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Shape()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_shape.py:44]: Can you revert this file to the master branch version?

File: samples/openapi3/client/petstore/python-experimental/test/test_triangle.py [hunk 17]
```diff
@@ -44,12 +44,9 @@ def tearDown(self):
...
         """Test Triangle"""
-        tri = Triangle(shape_type="Triangle", triangle_type="EquilateralTriangle")
-        assert isinstance(tri, equilateral_triangle.EquilateralTriangle)
-        tri = Triangle(shape_type="Triangle", triangle_type="IsoscelesTriangle")
-        assert isinstance(tri, isosceles_triangle.IsoscelesTriangle)
-        tri = Triangle(shape_type="Triangle", triangle_type="ScaleneTriangle")
-        assert isinstance(tri, scalene_triangle.ScaleneTriangle)
+        # FIXME: construct object with mandatory attributes with example values
+        # model = Triangle()  # noqa: E501
+        pass
```
Comment (spacether) [inline samples/openapi3/client/petstore/python-experimental/test/test_triangle.py:49]: Can you revert this file to the master branch version?

File: modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml (hunk 5 above)
Comment (jirikuncar) [reply modules/openapi-generator/src/test/resources/3_0/python-experimental/petstore-with-fake-endpoints-models-for-testing-with-http-signature.yaml:1103]: I would prefer to have it in separate API.

File: samples/openapi3/client/petstore/python-experimental/test/test_animal_farm.py (hunk 7 above)
Comment (jirikuncar) [reply samples/openapi3/client/petstore/python-experimental/test/test_animal_farm.py:39]: @spacether I will go back to my fake example. There are too many issues with `Animal` models.

File: samples/openapi3/client/petstore/python-experimental/tests/test_api_validation.py [hunk 18]
```diff
@@ -42,6 +43,12 @@ def test_configuration(self):
             config.disabled_client_side_validations = 'foo'
         config.disabled_client_side_validations = ""
 
+    def test_array_of_enums(self):
+        data = [
+            "placed", None
+        ]
+        response = MockResponse(data=json.dumps(data))
+        deserialized = self.api_client.deserialize(response, (array_of_enums.ArrayOfEnums, ), True)
```
Comment (jirikuncar) [inline samples/openapi3/client/petstore/python-experimental/tests/test_api_validation.py:51]: @spacether this does not work. I am going with following hack meantime:
```suggestion
        deserialized = self.api_client.deserialize(response, (list, ), True)
```

**Comment by jirikuncar** [comment]
@spacether what is the intended use for `Model(Simple|Normal)` with lists?

File: samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md (hunk 4 above)
Comment (jirikuncar) [reply samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md:5]: The problem is that `{{#vars}}` is empty for arrays.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java [hunk 19]
```diff
@@ -843,7 +841,20 @@ public CodegenModel fromModel(String name, Schema schema) {
                         postProcessModelProperty(result, prop);
                     }
                 }
+            } else if (result.isArrayModel) {
+                LOGGER.warn("\n\n{}\n\n", modelSchema);
+                LOGGER.warn("\n\n{}\n\n", modelProperty);
```
Comment (jirikuncar) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:847]:
```suggestion
```

File: samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md (hunk 4 above)
Comment (jirikuncar) [reply samples/openapi3/client/petstore/python-experimental/docs/AnimalFarm.md:5]: @spacether solved - see https://github.com/OpenAPITools/openapi-generator/pull/6493/files#diff-ec679c7ef213544fd9e5fdad21ffd6eeR6

**Comment by jirikuncar** [comment]
@spacether it should all be fixed now. I have run the sample generator locally but it is failing on CI. I don't know why 😞

**Comment by jimschubert** [comment]
@spacether looks like all your comments have been addressed here, is this ok to merge?
We can fix merge conflicts and regenerate via separate PR if you have no further concerns.

**Comment by spacether** [comment]
I will need a couple of days to dig in to this. It looks like the PR broke data types of unrelated samples so no I don't think that we should merge yet.

**Comment by jirikuncar** [comment]
> PR broke data types of unrelated samples

@spacether can you be more specific? The changes for additionalProperties seem to come from master. I will try to regenerate samples again.

**Comment by spacether** [comment]
The changes that I mentioned are the additionalProperties ones. If they came from master wouldn't rebasing on master make them disappear?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java [hunk 20]
```diff
@@ -829,13 +827,13 @@ public CodegenModel fromModel(String name, Schema schema) {
...
         // make non-object type models have one property so we can use it to store enums and validations
-        if (result.isAlias || result.isEnum) {
+        if (result.isAlias || result.isEnum || result.isArrayModel) {
             Schema modelSchema = ModelUtils.getSchema(this.openAPI, result.name);
             CodegenProperty modelProperty = fromProperty("value", modelSchema);
-            if (modelProperty.isEnum == true || modelProperty.hasValidation == true) {
+            if (modelProperty.isEnum == true || modelProperty.hasValidation == true || result.isArrayModel) {
                 // these models are non-object models with enums and/or validations
                 // add a single property to the model so we can have a way to access validations
-                result.isAlias = true;
```
Comment (spacether) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:838]: Array models are making a model that aliases the array type, so I think that this should be true.
Long term my goal is to have the ModelSimple classes inherrit from their base classes like str int etc.
Why do we need to set this to false for Arraymodels? Historically generateAliasAsModel applied only to Map and Array models in most generators. In python-experimental we apply it to those and all primitives only if they have validations and ebums. Standard processing in the generators was to describe all these primitive models as their simple types like str, list. But if those models had validations and enums that data was lost in that process. So I added code to preserve that information and make object type models which preserve the validation information which would have been lost.

After I did that CodegenModel was updated to include validations which previously had been missing. So in summary this code here is an older hack to preserve information. If we can use the existing ArrayModels as is that is preferable. It should define its datatype somewhere inside the CodegebModel and that should be used to set the ModelSimple data type in the python class for our made up calue parameter.

We can do this by doing a check of the model type when we assign the openapi types in ModelSimple.
We can check the data that is available by running the generator with a debugmodels flag.

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java [hunk 21]
```diff
@@ -270,7 +270,7 @@ public void arrayModelTest() {
         Assert.assertEquals(cm.classname, "sample.Sample");
         Assert.assertEquals(cm.classVarName, "sample");
         Assert.assertEquals(cm.description, "an array model");
-        Assert.assertEquals(cm.vars.size(), 0);
+        Assert.assertEquals(cm.vars.size(), 1);  // there is one value for Childer definition
```
Comment (spacether) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java:273]: What is the dataType for this sample?
It having vars of length 1 shows that my older hack is being used and that we have hoisted model information into a new variable.
How about we keep the model as is an use the information from
CodegenModel.dataType or CodegenModel.arrayModelType?
We can see all of the properties that are available by looking at that class [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java#L29)

We can debug the CodegenModels and see the information in them by running the generate command with the `-DdebugModels=true` as [shown here](https://github.com/openapitools/openapi-generator/wiki/FAQ#how-to-debug-openapi-generator)

How about in the classvars template here
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/classvars.mustache#L115
Doing a check to see if the model is an arrayModel
and setting the datatype accordingly?
It would look like:
```
{{#isArrayModel}}
            'value': ({{{dataType}}},),  # noqa: E501
{{/isArrayModel}}
{{^isArrayModel}}
{{#requiredVars}}
            '{{name}}': ({{{dataType}}},),  # noqa: E501
{{/requiredVars}}
{{#optionalVars}}
            '{{name}}': ({{{dataType}}},),  # noqa: E501
{{/optionalVars}}
{{/isArrayModel}}
```
We would also need to do a similar change in 
[method_init_shared](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/method_init_shared.mustache#L2)
To:
- not loop over required an optional args if we have an array model, if we are that case, just set value as a required positional arg
- else use our existing looping over variables
- in the docstring if we are an arraymodel set value as a required positional arg

@jirikuncar 
What do you think about making these changes?
If you don't want to make them, we can keep using the existing code/(my older solution) where we shovel the model data into a value variable.

**Comment by jirikuncar** [comment]
> The changes that I mentioned are the additionalProperties ones. If they came from master wouldn't rebasing on master make them disappear?

It should but there is some problem with the sample generation. I will continue updating my branch until it resolves.

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java (hunk 21 above)
Comment (jirikuncar) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java:273]: I would prefer to keep the current code. It looks like too many changes and `{{#isArrayModel}}` guards are required for templates.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java (hunk 20 above)
Comment (jirikuncar) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:838]: I am not aware of all the corner cases in the code. I was trying to make it work for my use-case and not break existing tests. Feel free to propose code changes.

Comment (jirikuncar) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java:838]: I hope I have found a sufficient fix in b2ca2f78e295ce1d94b42f6ebc55d9d5a64f1918.

**Comment by jirikuncar** [comment]
@jimschubert I have run the `./bin/openapi3/python-experimental-petstore.sh` and also tried `./bin/utils/ensure-up-to-date --batch` but the generated code for additional properties is still there.

**Comment by jirikuncar** [comment]
@spacether can we go with this solution for a time being?
@jimschubert the CI is passing 🎉

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java (hunk 21 above)
Comment (spacether) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/python/PythonClientExperimentalTest.java:273]: Ah, understood. That's fine.
